docs/source/extending/Queries.rst

```diff
@@ -0,0 +1,755 @@
+
+
+You can also use *macros*, which work like parameters but are substituted
+into the query before any parameter substitution is carried out. This
+allows you to, for example, use a single base query but with different
+filter and summarization clauses defined as macros. The macro text is
+substituted into the main query.
+
+Macros are added to the ``query_macros`` subkey of a query. They have
+two subkeys: description and value. value defines the text to be inserted.
+The key name is the name of the macro.
+
+In the query, you denote the substitution point by surrounding the macro name
+with "$<" and ">$". This is show in the example below.
+
+.. code:: yaml
+
+    - query: '
+        {table}
+        | where SubType_s == "FlowLog"
+        | where FlowStartTime_t >= datetime({start})
+        | where FlowEndTime_t <= datetime({end})
+        $<query_condition>$
+        | where (AllowedOutFlows_d > 0 or AllowedInFlows_d > 0)
+        {query_project}
+        | extend AllExtIPs = iif(isempty(PublicIPs), pack_array(ExtIP),
+                         iif(isempty(ExtIP), PublicIPs, array_concat(PublicIPs, pack_array(ExtIP)))
+                         )
+        | project-away ExtIP
+        | mvexpand AllExtIPs
+        {add_query_items}'
+
+Macros are particularly useful when combined with yaml aliases. You can, for
+example, define a base query (using a yaml alias) with a macro reference in the
+query body. Then in each query definition you can have different macro values
+for the macro to be substituted. For example:
+
+.. code:: yaml
+
+    metadata:
+        ...
+        aliases:
+            - &azure_network_base_query '
+                {table}
+                | where SubType_s == "FlowLog"
+                | where FlowStartTime_t >= datetime({start})
+                | where FlowEndTime_t <= datetime({end})
+                $<query_condition>$
+                | where (AllowedOutFlows_d > 0 or AllowedInFlows_d > 0)
+                {query_project}
+                | extend AllExtIPs = iif(isempty(PublicIPs), pack_array(ExtIP),
+                                iif(isempty(ExtIP), PublicIPs, array_concat(PublicIPs, pack_array(ExtIP)))
+                                )
+                | project-away ExtIP
+                | mvexpand AllExtIPs
+                {add_query_items}'
+        ...
+    sources:
+        list_azure_network_flows_by_ip:
+            description: Retrieves Azure network analytics flow events.
+        args:
+            query: *azure_network_base_query
+        parameters:
+            query_project:
+                ...
+            end:
+                description: Query end time
+                type: datetime
+        query_macros:
+            query_condition:
+                description: Query-specific where clause
+                value: '| where (VMIP_s in ({ip_address_list})
+                or SrcIP_s in ({ip_address_list})
+                or DestIP_s in ({ip_address_list})
+                )'
```

Comment:
i'm not sure this example makes it clear enough why someone would use the macro here.